Refactor AST/Environment

markhalliwell opened the pull request:
@colinodell this is more or less what I was thinking. It would be nice if there were a common namespace that both projects could inherit from, that way other libraries could pass along their own `Node` objects. As it stands now, I think this may be the direction I want to go in, cause it would allow us to create a `TwemojiExtension` that adds a custom renderer for all `\UnicornFail\Emoji\Node\Inline\Unicode` objects.

markhalliwell committed:
WIP: Basic implementation of commonmark’s AST & Environment code

Signed-off-by: Mark Carver <mark.carver@me.com>

colinodell commented on src/Node/Node.php:
```diff
@@ -0,0 +1,263 @@
+     *
+     * @psalm-readonly-allow-private-mutation
+     */
+    protected $parent;
```
I think the reuse of commonmark's full AST implementation might be overkill here.  In commonmark, we need a way to represent a **nested tree** of parsed objects, and that representation needed to be **implemented as a multi-level, doubly-linked list** so we can iterate it without using recursion.  But in this case, I think we could get by with a **flat, non-tree** representation since we're dealing with a stream of characters.  This could be a linked list, doubly-linked list, or even just a simple `iterable` of nodes/tokens. I just don't think we really need the concept of parents and children.

colinodell replied:
And if we're making it flat I also think we can remove the concept of block vs inline nodes.  The object that contains the nodes doesn't necessarily have to be a node itself in this case.

markhalliwell replied:
I agree that the "nested/depth" aspect of this isn't necessary here; this was just a literal c&p for times sake.

I mostly extracted it because I would like to keep the node traversal and injection methods (i.e. append, detach, insertBefore/After, previous, next, etc.). These kinds of methods make it easier to manipulate the AST, regardless of its depth.

I do think there should be a simple (flat) `NodeInterface` and then a `HierarchicalNodeInterface` perhaps? Again, the main goal here is to see what aspects could be extracted out that aren't really "commonmark" specific and could be reused between the two projects and provide a common interface for both to use so what is parsed here could be used "as is" in commonmark without having to do unnecessary and (ultimately) costly object transforms.

markhalliwell replied:
This has been simplified/removed.

colinodell commented on src/Input/Input.php:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @psalm-readonly-allow-private-mutation
+     */
+    private $lines;
```
I don't think you necessarily need the `src/Input` stuff from commonmark.  We need it there because implementing the spec requires us to consider **each line** as a whole since the parsing behavior might change based on what the previous/next line contains.

In this case, I don't think we necessarily need to consider lines since we're primarily working at the character-level - I don't think `\n` has any special meaning, at least not one that would be different from other whitespace characters.

markhalliwell replied:
I think this is needed, if only from a performance level (memory usage). I could be mistaken, but it seems like it would make more sense if `preg_split` worked on individual lines, rather than the document as a whole.

colinodell commented on src/Output/RenderedContentInterface.php:
```diff
@@ -0,0 +1,29 @@
+
+use UnicornFail\Emoji\Node\Block\Document;
+
+interface RenderedContentInterface extends \Stringable
```
We needed this in commonmark as a workaround to provide additional metadata about fully-parsed-and-rendered content - specifically the front matter.  We create a new instance of an object that implements this interface but also provides a method to grab the parsed front matter.  (I'm not happy with that implementation and might be removing it completely.)

For this library, I think we can get away with the rendering step outputting a simple string.

markhalliwell replied:
Perhaps. I honestly prefer passing around objects that implement interfaces rather than dealing with scalars if at all possible; especially in any sort of "extensible" environment. Given that there is `\Stringable`, it makes it easy enough to convert into a string when it ultimately needs to be. Premature rendering can often be an issue.

markhalliwell replied:
This has been simplified/removed.

colinodell commented on src/Parser/Parser.php:
```diff
@@ -0,0 +1,141 @@
+use UnicornFail\Emoji\Node\Inline\Text;
+use UnicornFail\Emoji\Node\Inline\Unicode;
+
+class Parser implements ParserInterface
```
I haven't dug deeply into the parsing/lexing implementation yet so I don't have any comments here yet, other than that representing the parsed tokens as a flat "stream" (where each item in that "stream" represents either a single character or a group of characters) aligns very nicely with how the lexing works and how parsing UTF-8 words.  I therefore think we can keep the abstraction fairly straight-forward.

markhalliwell replied:
This has been simplified and the `Input` class was removed.

colinodell commented on src/Renderer/Inline/TextRenderer.php:
```diff
@@ -0,0 +1,37 @@
+    public function render(Node $node, ChildNodeRendererInterface $childRenderer)
+    {
+        if (! ($node instanceof Text)) {
+            throw new \InvalidArgumentException('Incompatible node type: ' . \get_class($node));
```
Will there only be a finite number of node types?  In other words, do we ever plan to support more than just emojis, shortcodes, emoticons, and whatever else you might have implemented?

If the types are finite and new types cannot be added by users then we might be able to simplify the rendering process in a more type-strict way.  I don't have any specific ideas or suggestions at the moment - just curious right now :)

markhalliwell replied:
> Will there only be a finite number of node types?

At this moment in time, I do not know. I really don't want to box the project in though and would like to leave this open to interpretation. My goal is to get away from switch/if statements based on the lexer "type" and would rather let a node indicate whether it matches. I'm still not 💯 how I want to do this.

The primary goal of this though was to show the logical workflow as to what would be considered a "node". If some aspects of commonmark were extracted into a simpler library that both of these projects could share (i.e. `league/ast` ) or at the very least some sort of rudimentary interface that could be shared, then these nodes could be passed directly to commonmark as is (there wouldn't be any need to "transform them into commonmark nodes").

markhalliwell replied:
With extensions, yes, this can be any kind of Node (i.e. images)

colinodell reviewed (commented):
I won't be able to review this in-depth until Friday or Saturday at the earliest, but I do have some initial, high-level thoughts.

markhalliwell commented on src/Environment/AbstractConfigurableEnvironment.php:
```diff
@@ -0,0 +1,31 @@
+use UnicornFail\Emoji\Configuration\ConfigurationAwareInterface;
+use UnicornFail\Emoji\Traits\ConfigurableEnvironmentTrait;
+
+class AbstractConfigurableEnvironment extends AbstractEnvironment implements ConfigurableEnvironmentInterface
```
All of these "Environment" based abstractions are intended to go into something like `league/environment` or even part of `league/ast` or have the latter provide additional "renderable" aspects.

The goal was to split out the environment components into separate interfaces and traits, so it could be "built-up" as needed. Still not 💯 there, but I hope this gives an idea of what I'm after and what could be easier to implement in the long run for both projects.

markhalliwell replied:
This has been simplified to use `league/configuration` (https://github.com/unicorn-fail/configuration)

markhalliwell commented on src/Environment/Environment.php:
```diff
@@ -0,0 +1,112 @@
+use UnicornFail\Emoji\Parser\Parser;
+use UnicornFail\Emoji\Parser\ParserInterface;
+
+final class Environment extends AbstractConfigurableEnvironment implements EmojiEnvironmentInterface
```
Clearly commonmark will have very specific requirements, but that can easily be addressed in a `CommonMarkEnvironmentInterface` as I did here with `EmojiEnvironmentInterface`.

markhalliwell replied:
This has been simplified.

markhalliwell commented:
@colinodell any further thoughts on this?

colinodell commented:
> @colinodell any further thoughts on this?

My apologies!  I still owe you a review and feedback on everything else here - I'll try to get that done this weekend. I appreciate the ping!

colinodell commented:
I'll give you some initial, high-level feedback now :)

At this point, I don't have a strong desire to split parts of `league/commonmark` into separate packages.  The only parts I'd be somewhat inclined to split out would be the configuration and event systems, as they are very general-purpose.  I don't envision splitting the AST or environment-specific code out as most of that is tightly coupled to the league/commonmark engine.

In terms of parsing input, I'm thinking about this mostly in the context of how it would fit integrate with `league/commonmark`.  In my mind there are two possible ways you might parse emoji:

 - Making a custom inline parser to handle things like `:alias:` shortcode input
 - Making a custom event observer to check for emoji characters or emoticons inside of parsed `Text` elements

In the first case, your parser tells the engine to stop at any `:` - you then try to parse the next few characters as an alias.  In the second case, you're looking for any Unicode characters that are valid emoji or emoticons that match a specific sequence of characters.

I believe the following hold true for both cases:

 - You don't need to worry about splitting things into lines - if there's a `\n` character somewhere it simply fails to match either `:alias:` or an emoji character. Therefore, `\n` behaves just like any other character and doesn't need special treatment.
 - If you find an `:alias:`, emoji, or emoticon you'll be splitting that out into an `Emoji` node of some kind.  Everything else will be left as `Text`.  Therefore you don't need a nested AST implementation, just the ability to split strings into `Emoji` or `Text` and put those into some kind of flat, linear list.

For those reasons, I don't think your input parsing approach or AST has to align with `league/commonmark`.  You can still use some kind of AST and/or node types internally, though.  `Emoji` nodes in a `league/commonmark` extension could still directly embed objects from this library in the nodes' properties if desired.

I have not reviewed the environment stuff here, and I'm sure I missed a few points you mentioned above, so I'll circle back to those (and any other thoughts you have) later this week.

markhalliwell commented:
A lot of this work was based on the previous conversations we had about extracting non-commonmark specific code. Things, like environment and config, are certainly not commonmark specific.

In any case, if there's no interest in sharing code then I can certainly simplify it to be less abstract here. My goal was only to stop having to constantly create many of these common requirements. In hindsight and given the amount of time that has passed, it probably makes more sense to (at least now) to not try and change everything.

Re: parsing/line splitting, the primary goal of this particular project is to be a standalone implementation for parsing and converting emojis, emoticons and shortcodes. This has nothing really to do with commonmark and users may need/want to parse a multiline document on its own.

While we could parse things in commonmark at a character level, I believe it's just easier to register `DocumentParsedEvent` in commonmark and then walk over all the text elements found (as https://github.com/thephpleague/commonmark/pull/546 shows). Given the various types of emojis, emoticons and shortcodes that need to be supported, commonmark doesn't need to know how to parse these; just integrate with this implementation and let it do its own thing.

markhalliwell commented:
I've gone ahead and extracted the configuration stuff into its own repo/project and added you as an admin:

https://github.com/unicorn-fail/configuration

Once you deem that it's ready, it should be relatively easy to transfer into the `league` namespace.

markhalliwell committed:
Replace configuration with league/configuration

markhalliwell committed:
Start to simplify environment and add twemoji extension for real use case

markhalliwell commented:
I'm starting to pair down the environment stuff and went ahead and added a Twemoji extension so we can use a real use case scenario to help build it out to determine what is actually needed. Needless to say, this is still a WIP, so I'll ping you once its ready again so you don't accidentally waste your time reviewing.

markhalliwell committed:
Simplify Document and Node, remove unnecessary Input, NodeWalker, RenderedContent

markhalliwell commented:
All tests/coverage is 100% again locally. Need to figure out what's going on with GH actions. I think it may be due to a PHP version/serialization discrepancy (pretty sure the archives need to be [serialized in PHP 7.2 as things changed in 7.4](https://github.com/unicorn-fail/emoji/blob/f8f86629b75b24aa72e31057587eb55f3b787503/tests/unit/Dataset/EmojiTest.php#L424-L429)).

@colinodell I know this is a lot of code to review, so I'm thinking I'm going to go ahead and merge it and we can follow-up with smaller issues/PRs moving forward?

edit: I re-read a lot of your comments and made adjustments accordingly to simplify and "flatten" the AST. There's still a lot of boiler code, but that's mainly due to the necessity for extensions (Twemoji).

markhalliwell committed:
Rebuild dataset archives and fix composer.lock to work with PHP 7.2+
